From 09c0eda72510258eeef01ffd05dc8fa15f67f872 Mon Sep 17 00:00:00 2001
From: Monika Kairaityte <monika@kibit.lt>
Date: Mon, 16 Nov 2020 12:27:58 +0200
Subject: [PATCH] try: Fix bytes-related master crash when calling buildbot try

---
@@ -0,0 +1 @@
+Fix bytes-related master crash when calling buildbot try (:issue:`4488`)
diff --git a/usr/lib/python3.9/site-packages/buildbot/schedulers/trysched.py b/usr/lib/python3.9/site-packages/buildbot/schedulers/trysched.py
index 86f3c48fc..91f801407 100644
--- a/usr/lib/python3.9/site-packages/buildbot/schedulers/trysched.py
+++ b/usr/lib/python3.9/site-packages/buildbot/schedulers/trysched.py
@@ -393,6 +393,14 @@ class Try_Userpass_Perspective(pbutil.NewCredPerspective):
         if not builderNames:
             return None
 
+        branch = bytes2unicode(branch)
+        revision = bytes2unicode(revision)
+        patch = patch[0], bytes2unicode(patch[1])
+        repository = bytes2unicode(repository)
+        project = bytes2unicode(project)
+        who = bytes2unicode(who)
+        comment = bytes2unicode(comment)
+
         reason = "'try' job"
 
         if who:
diff --git a/usr/lib/python3.9/site-packages/buildbot/test/integration/test_try_client.py b/usr/lib/python3.9/site-packages/buildbot/test/integration/test_try_client.py
index ce703db1c..a21bfc400 100644
--- a/usr/lib/python3.9/site-packages/buildbot/test/integration/test_try_client.py
+++ b/usr/lib/python3.9/site-packages/buildbot/test/integration/test_try_client.py
@@ -66,10 +66,10 @@ class Schedulers(RunMasterBase, www.RequiresWwwMixin):
 
         self.patch(reactor, 'spawnProcess', spawnProcess)
 
+        self.sourcestamp = tryclient.SourceStamp(branch='br', revision='rr', patch=(0, '++--'))
+
         def getSourceStamp(vctype, treetop, branch=None, repository=None):
-            return defer.succeed(
-                tryclient.SourceStamp(branch='br', revision='rr',
-                                      patch=(0, '++--')))
+            return defer.succeed(self.sourcestamp)
         self.patch(tryclient, 'getSourceStamp', getSourceStamp)
 
         self.output = []
@@ -160,6 +160,30 @@ class Schedulers(RunMasterBase, www.RequiresWwwMixin):
         buildsets = yield self.master.db.buildsets.getBuildsets()
         self.assertEqual(len(buildsets), 1)
 
+    @defer.inlineCallbacks
+    def test_userpass_wait_bytes(self):
+        self.sourcestamp = tryclient.SourceStamp(branch=b'br', revision=b'rr', patch=(0, b'++--'))
+
+        yield self.startMaster(
+            trysched.Try_Userpass('try', ['a'], 0, [('u', b'p')]))
+        yield self.runClient({
+            'connect': 'pb',
+            'master': '127.0.0.1:{}'.format(self.serverPort),
+            'username': 'u',
+            'passwd': b'p',
+            'wait': True,
+        })
+        self.assertEqual(self.output, [
+            "using 'pb' connect method",
+            'job created',
+            'Delivering job; comment= None',
+            'job has been delivered',
+            'All Builds Complete',
+            'a: success (build successful)',
+        ])
+        buildsets = yield self.master.db.buildsets.getBuildsets()
+        self.assertEqual(len(buildsets), 1)
+
     @defer.inlineCallbacks
     def test_userpass_wait_dryrun(self):
         yield self.startMaster(
-- 
2.26.2

